File: tests/r_parity/73_did2s.py
"""StatsPAI Gardner two-stage DiD parity (Python side) -- Module 73.

Runs ``sp.gardner_did`` on the mpdta replica; the companion R script runs
``did2s::did2s`` with the same two-way FE first stage on the same CSV, and
the Stata do-file runs Butts' ``did2s`` port.

Tolerance: rel < 1e-6 on the point estimate and on the SE. The default
``vce='analytic'`` is the did2s corrected clustered variance (Gardner 2022):
the Stage-2 sandwich is built from the two-stage influence function, so the
Stage-1 fixed-effect estimation error is propagated exactly as the two
references do, with no small-sample cluster factor. Before this correction
the default clustered the stage-2 residuals only and landed ~26% low; that
legacy number is still reachable as ``vce='stage2'`` and is emitted here as
the ``static_ATT_stage2_se`` diagnostic row (no R/Stata counterpart, not
joined by compare.py) so the size of the old gap stays on record.
"""

from __future__ import annotations

import warnings

import statspai as sp

from _common import ParityRecord, dump_csv, write_results

MODULE = "73_did2s"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    fit = sp.gardner_did(
        df,
        y="lemp",
        group="countyreal",
        time="year",
        first_treat="first_treat",
    )
    with warnings.catch_warnings():
        # vce='stage2' warns that it understates; that is the legacy
        # convention this diagnostic row exists to record.
        warnings.simplefilter("ignore")
        legacy = sp.gardner_did(
            df,
            y="lemp",
            group="countyreal",
            time="year",
            first_treat="first_treat",
            vce="stage2",
        )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="static_ATT",
            estimate=float(fit.estimate),
            se=float(fit.se),
            ci_lo=float(fit.ci[0]) if fit.ci is not None else None,
            ci_hi=float(fit.ci[1]) if fit.ci is not None else None,
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="static_ATT_stage2_se",
            estimate=float(legacy.estimate),
            se=float(legacy.se),
            n=int(len(df)),
        ),
    ]

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "estimator": "gardner_did",
            "vce": "analytic (did2s corrected two-stage variance) + stage2 diagnostic",
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/73_did2s.R
# StatsPAI Gardner two-stage DiD parity (R side) -- Module 73.
#
# Reads data/73_did2s.csv (the StatsPAI mpdta replica) and runs
# did2s::did2s with a two-way FE first stage and a static treatment
# indicator in the second stage.
#
# Tolerance: rel < 1e-6 on the point estimate and on the SE. did2s reports
# the corrected clustered variance of Gardner (2022), which propagates
# first-stage estimation error into the second-stage sandwich with no
# small-sample cluster factor; sp.gardner_did's default vce='analytic'
# builds the same two-stage influence function, so the SE row is a strict
# parity assertion (observed rel 2.7e-10; the residual is fixest's
# iterative-demeaning tolerance in the first stage, which also shows in
# the 4.8e-8 point-estimate gap -- Stata's exact first stage lands on the
# Python SE at 1e-14).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(did2s)
})

MODULE <- "73_did2s"

df <- read_csv_strict(MODULE)
df$first_treat <- as.numeric(df$first_treat)
# Time-varying treatment status implied by the cohort column.
df$dpost <- as.integer(df$first_treat > 0 & df$year >= df$first_treat)

fit <- did2s::did2s(
  df,
  yname        = "lemp",
  first_stage  = ~ 0 | countyreal + year,
  second_stage = ~ i(dpost, ref = FALSE),
  treatment    = "dpost",
  cluster_var  = "countyreal"
)

co <- coef(fit)
se <- sqrt(diag(vcov(fit)))

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "static_ATT",
    estimate  = unname(co[1]),
    se        = unname(se[1]),
    n         = nrow(df)
  )
)

write_results(MODULE, rows, extra = list(
  package = "did2s",
  version = as.character(utils::packageVersion("did2s"))
))
